File: sdks/python3/nanoservice/node/node_handler.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Any, Dict, Tuple

from nanoservice.types.context import Context


class NodeHandler(ABC):
    """Base class that all Blok nodes must inherit from.

    Subclasses must implement the ``execute`` method which receives the
    workflow context and node-specific configuration, and returns a
    data value (any JSON-serializable value).

    Example::

        class GreetNode(NodeHandler):
            def execute(self, ctx, config):
                name = ctx.request.body_str("name") or "World"
                prefix = config.get("prefix", "Hello")
                return {"message": f"{prefix}, {name}!"}
    """

    @abstractmethod
    def execute(self, ctx: Context, config: Dict[str, Any]) -> Any:
        """Execute the node logic.

        Args:
            ctx: The workflow execution context.
            config: Node-specific configuration.

        Returns:
            JSON-serializable result data.
        """
        raise NotImplementedError

File: sdks/ruby/lib/nanoservice/node/node_handler.rb
# frozen_string_literal: true

module Nanoservice
  module Node
    # NodeHandler is the base class that all Blok nodes must inherit from.
    #
    # Subclasses must implement the +execute+ method which receives the
    # workflow context and node-specific configuration, and returns a
    # data hash (or any JSON-serializable value).
    #
    # @example
    #   class GreetNode < Nanoservice::Node::NodeHandler
    #     def execute(ctx, config)
    #       name = ctx.request.body_str("name") || "World"
    #       prefix = config["prefix"] || "Hello"
    #       { "message" => "#{prefix}, #{name}!" }
    #     end
    #   end
    #
    class NodeHandler
      # Execute the node logic.
      #
      # @param ctx [Nanoservice::Types::Context] The workflow execution context
      # @param config [Hash] Node-specific configuration
      # @return [Object] JSON-serializable result data
      # @raise [NotImplementedError] if not overridden by a subclass
      def execute(ctx, config)
        raise NotImplementedError, "#{self.class}#execute must be implemented"
      end
    end
  end
end
